src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs

```diff
@@ -249,5 +249,456 @@ public static void Main()
+                //     readonly ~S() { }
+                Diagnostic(ErrorCode.ERR_BadMemberFlag, "S").WithArguments("readonly").WithLocation(4, 15),
+                // (4,15): error CS0575: Only class types can contain destructors
+                //     readonly ~S() { }
+                Diagnostic(ErrorCode.ERR_OnlyClassesCanContainDestructors, "S").WithArguments("S.~S()").WithLocation(4, 15));
+        }
+
+        [Fact]
+        public void ReadOnlyOperator()
+        {
+            var csharp = @"
+public struct S
+{
+    public static readonly S operator +(S lhs, S rhs) => lhs;
+    public static readonly explicit operator int(S s) => 42;
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics(
+                // (4,39): error CS0106: The modifier 'readonly' is not valid for this item
+                //     public static readonly S operator +(S lhs, S rhs) => lhs;
+                Diagnostic(ErrorCode.ERR_BadMemberFlag, "+").WithArguments("readonly").WithLocation(4, 39),
+                // (5,46): error CS0106: The modifier 'readonly' is not valid for this item
+                //     public static readonly explicit operator int(S s) => 42;
+                Diagnostic(ErrorCode.ERR_BadMemberFlag, "int").WithArguments("readonly").WithLocation(5, 46));
+        }
+
+        [Fact]
+        public void ReadOnlyDelegate()
+        {
+            var csharp = @"
+public readonly delegate int Del();
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics(
+                // (2,30): error CS0106: The modifier 'readonly' is not valid for this item
+                // public readonly delegate int Del();
+                Diagnostic(ErrorCode.ERR_BadMemberFlag, "Del").WithArguments("readonly").WithLocation(2, 30));
+        }
+
+        [Fact]
+        public void ReadOnlyIndexer()
+        {
+            var csharp = @"
+public struct S1
+{
+    public readonly int this[int i]
+    {
+        get => 42;
+        set {}
+    }
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ReadOnlyExpressionIndexer()
+        {
+            var csharp = @"
+public struct S1
+{
+    public readonly int this[int i] => 42;
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ReadOnlyGetExpressionIndexer()
+        {
+            var csharp = @"
+public struct S1
+{
+    public int this[int i]
+    {
+        readonly get => 42;
+        readonly set {}
+    }
+}
+";
+            var comp = CreateCompilation(csharp);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ReadOnlyFieldLikeEvent()
+        {
+            var csharp = @"
+using System;
+
+public struct S1
+{
+    public readonly event Action<EventArgs> OnFoo;
```

Comment:
>OnFoo [](start = 44, length = 5)

Please use another name. Perhaps `E`. Same comment for tests below.

Reply:
Done.